# **01. Exploracion Inicial del Datase.**


|**Proyecto** | Violencia Intrafamiliar en Colombia - Policia Nacional |
|---          |---|
|**Objetivo** | Auditar la calidad del dataset crudo sin modificar ningun valor |
|**Dataset**  | `data/raw/Reporte_Delito_Violencia_Intrafamiliar_Policia_Nacional.csv` |
|**Autor**    | Ovidio Antonio Guerrero Mosquera |
|**Fecha**    | 2026-04-24 |

---

### Contexto:
Este notebook es el punto de partida del analisis. Su unico proposito es **conocer y diagnosticar** el dataset tal cual como fue entregado por la fuente. No se modifica ningun valor ni se hara transformaciones, eso se hara en el notebook 02.


### Preguntas que guiaran esta exploracion:

1. ¿Cual es la estrucutra del dataset (filas, columnas, tipos de datos)?
2. ¿Existen valores nulos?¿En que columnas y en que magnitud?
3. ¿Existen filas duplicadas exactas?
4. ¿Los valores unicos de cada columna categorica son coherentes?
5. ¿La variable `CANTIDAD` tiene valores atipicos?
6. ¿El rango temporal es consistente con lo esperado?

## 0. Importaciones y Configuraciones:

In [76]:
import pandas as pd
import numpy as np

# Opciones de Pantalla de Pandas.
pd.set_option('display.max_columns', None)
# pd.set_option('display.float_format', '{:.2f}'.format)

print("Librerias cargadas exitosamente ✅!!!")

Librerias cargadas exitosamente ✅!!!


## 1. Carga del Dataset:

Se carga el CSV crudo desde `data/raw` sin modificar ningun valor. `low_memory=False` para evitar advertencias por columnas con tipo de datos mixtos.

In [77]:
ruta = "../data/raw/Reporte_Delito_Violencia_Intrafamiliar_Policia_Nacional.csv"

df = pd.read_csv(ruta,
                 low_memory=False,
                 sep=',',
                 encoding='utf-8',
)

print("El Dataset se cargo exitosamente ✅!!!")

El Dataset se cargo exitosamente ✅!!!


## 2. Inspeccion General:

Se revisa las dimensiones del Dataset, nombre de las columnas, tipos de datos, una muestra visual de los registros, resumen estadistica descriptiva, etc.

In [78]:
# Se obtienen las dimensiones del dataset.

print(f"{'='*40} DIMENSIONES DEL DATASET {'='*40}")

print(f"\nEl Datase tiene {df.shape[0]:,} filas y {df.shape[1]} columnas.")

======================================== DIMENSIONES DEL DATASET ========================================

El Datase tiene 660,756 filas y 8 columnas.


In [79]:
# Se obtinene los nombre de las columnas del DataFrame
print(f"{'='*40} NOMBRES DE LAS COLUMNAS {'='*40}")

df.columns.to_list()

======================================== NOMBRES DE LAS COLUMNAS ========================================


['DEPARTAMENTO',
 'MUNICIPIO',
 'CODIGO DANE',
 'ARMAS MEDIOS',
 'FECHA HECHO',
 'GENERO',
 'GRUPO ETARIO',
 'CANTIDAD']

**Hallazgos:**
* El dataset tiene `660,756 filas` y `8 columnas`.
* Todos los nombres de las columnas estan en mayuscula.
* Las columnas `CODIGO DANE`, `FECHA HECHO` y `GRUPO ETARIO` tinen espacios en sus nombres, lo que requerira atencion en el cuaderno `02_limpieza.ipynb`.

In [80]:
# Se obtiene el tipo de dato de cada una de las columnas.
print(f"{'='*40} TIPO DE DATO DE CADA UNA DE LAS COLUMNAS {'='*40}")

df.dtypes

======================================== TIPO DE DATO DE CADA UNA DE LAS COLUMNAS ========================================


DEPARTAMENTO      str
MUNICIPIO         str
CODIGO DANE     int64
ARMAS MEDIOS      str
FECHA HECHO       str
GENERO            str
GRUPO ETARIO      str
CANTIDAD        int64
dtype: object

**Hallazgo:**
* Las columnas `DEPARTAMENTO`, `MUNICIPIO`, `ARMAS MEDIOS`, `GENERO`, `GRUTO ETARIO` son tipo `str`.
* La columna `FECHA HECHO` es tipo `str`, no fue reconocida como fecha, por lo cual habra que convertirla a formato fecha en la etapaa de lilmpieza.
* La columna `CANTIDAD` es tipo `int64`.
* La columna `CODIGO DANE` es tipo `int64`, debera convertirse a `str`, dado que no aporta valor numerico.

In [81]:
# Se muestran las primeras 5 filas
print(f"{'='*40} PRIMERAS 5 FILAS DEL DATASET {'='*40}")

df.head()

======================================== PRIMERAS 5 FILAS DEL DATASET ========================================


,DEPARTAMENTO,MUNICIPIO,CODIGO DANE,ARMAS MEDIOS,FECHA HECHO,GENERO,GRUPO ETARIO,CANTIDAD
0,ANTIOQUIA,Medellín (CT),5001000,SIN EMPLEO DE ARMAS,18/09/2025,MASCULINO,MENORES,1
1,CUNDINAMARCA,Bogotá D.C. (CT),11001000,SIN EMPLEO DE ARMAS,23/10/2025,FEMENINO,MENORES,5
2,CUNDINAMARCA,Bogotá D.C. (CT),11001000,SIN EMPLEO DE ARMAS,23/10/2025,MASCULINO,MENORES,5
3,CAUCA,Cajibío,19130000,SIN EMPLEO DE ARMAS,23/10/2025,FEMENINO,ADULTOS,1
4,NORTE DE SANTANDER,Teorama,54800000,SIN EMPLEO DE ARMAS,06/10/2025,MASCULINO,ADULTOS,1


In [82]:
# Se muestran las ultimas 5 filas
print(f"{'='*40} ULTIMAS 5 FILAS DEL DATASET {'='*40}")

df.tail()

======================================== ULTIMAS 5 FILAS DEL DATASET ========================================


,DEPARTAMENTO,MUNICIPIO,CODIGO DANE,ARMAS MEDIOS,FECHA HECHO,GENERO,GRUPO ETARIO,CANTIDAD
660751,RISARALDA,Pereira (CT),66001000,ARMA BLANCA / CORTOPUNZANTE,14/06/2020,FEMENINO,ADULTOS,1
660752,QUINDÍO,Armenia (CT),63001000,ARMA BLANCA / CORTOPUNZANTE,16/06/2020,FEMENINO,ADULTOS,1
660753,ATLÁNTICO,Soledad,8758000,ARMA BLANCA / CORTOPUNZANTE,09/06/2020,MASCULINO,ADULTOS,1
660754,VALLE,Roldanillo,76622000,ARMA BLANCA / CORTOPUNZANTE,11/06/2020,FEMENINO,ADULTOS,1
660755,CUNDINAMARCA,Bogotá D.C. (CT),11001000,ARMA BLANCA / CORTOPUNZANTE,30/03/2020,FEMENINO,ADULTOS,1


In [83]:
# Se muestran 5 registros de manera aletoria para realizar una inspeccion inicial mas honesta
print(f"{'='*40} SELECCION ALEATORIA DE 5 FILAS DEL DATASET {'='*40}")

df.sample(5)

======================================== SELECCION ALEATORIA DE 5 FILAS DEL DATASET ========================================


,DEPARTAMENTO,MUNICIPIO,CODIGO DANE,ARMAS MEDIOS,FECHA HECHO,GENERO,GRUPO ETARIO,CANTIDAD
395194,HUILA,Pitalito,41551000,CONTUNDENTES,23/08/2014,FEMENINO,ADULTOS,1
632799,HUILA,La Plata,41396000,ARMA BLANCA / CORTOPUNZANTE,16/09/2012,FEMENINO,ADULTOS,1
536302,ATLÁNTICO,Barranquilla (CT),8001000,CONTUNDENTES,05/07/2020,FEMENINO,ADULTOS,4
437616,ATLÁNTICO,Soledad,8758000,CONTUNDENTES,16/02/2012,FEMENINO,ADULTOS,2
273638,CUNDINAMARCA,Soacha,25754000,NO REPORTADO,13/09/2020,MASCULINO,ADULTOS,1


**Hallazgo:**
* La columna `FECHA HECHO` tiene un formato consistente visible (DD/MM/AAAA).
* La columna `CANTIDAD` parece razonable en los registros visiables.
* En las primeras y ultimas 5 filas, al igual que en las 5 filas aleatorias, no se ven filas filas completamente vacias.
* Los textos en los valores de las columnas no estan estandarizados.

In [84]:
# Informacion del Dataset
print(f"{'='*40} INFORMACION DEL DATASET {'='*40}")

df.info()

======================================== INFORMACION DEL DATASET ========================================
<class 'pandas.DataFrame'>
RangeIndex: 660756 entries, 0 to 660755
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype
---  ------        --------------   -----
 0   DEPARTAMENTO  660756 non-null  str  
 1   MUNICIPIO     660752 non-null  str  
 2   CODIGO DANE   660756 non-null  int64
 3   ARMAS MEDIOS  660756 non-null  str  
 4   FECHA HECHO   660756 non-null  str  
 5   GENERO        658150 non-null  str  
 6   GRUPO ETARIO  659732 non-null  str  
 7   CANTIDAD      660756 non-null  int64
dtypes: int64(2), str(6)
memory usage: 40.3 MB


**Hallazgos:**

* Los nombres de todas las columnas estan escritas en mayuscula.

* Las columnas `ARMAS MEDIO`, `FECHA HECHO` Y `GRUPO ETARIO` contienen espacio en sus nombres.

* La columna `FECHA` esta tipada como `object`.

* El dataset contiene:
  * 2 variables numericas: `CODIGO DANE`, `CANTIDAD`.
  * 6 variables categoricas: `MUNICIPIO`, `DEPARTAMENTO`, `ARMAS MEDIO`, `FECHA HECHO`, `GRUPO ETARIO` y `GENERO`.

* La columna `CODIGO DANE` esta presente como variable numerica.

* Las columnas `MUNICIPIO`, `GENERO` Y `GRUPO ETARIO` presentan nulos.

In [85]:
# Resumen de estadistica descriptiva de las variables numericas
print(f"{'='*40} ESTADISTICA DESCRIPTIVA COLUMNAS NUMERICAS {'='*40}")

df.describe().round(2)

======================================== ESTADISTICA DESCRIPTIVA COLUMNAS NUMERICAS ========================================


,CODIGO DANE,CANTIDAD
count,660756.00,660756.00
mean,38701271.32,2.11
std,27324390.27,5.15
min,0.00,1.00
25%,13442000.00,1.00
50%,25781000.00,1.00
75%,68081000.00,2.00
max,99773004.00,164.00


**Hallazgo:**
* Hay 2 columnas numericas `CODIGO DANE` y `CANTIDAD`.
* La columna `CODIGO DANE` aunque es numerica no tiene sentido tratarla como numero, dado que no aporta ningun valor al analisis.
* El valor minimo de `CANTIDAD` es `1` y valor maximo es `164`, este ultimo es sospechoso dado que se trata de la cantidad de victimas del hecho de VIF.
* En la variable `CANTIDAD`: la `Media = 2.11`, la `Mediana = 1` y el valor `Maximo = 164`. Lo cual muestra una distribuccion fuertemente sesgada (cola hacia la derecha).

In [86]:
# Estadistica Descriptiva Columnas Categoricas

print(f"{'='*40} ESTADISTICA DESCRIPTIVA COLUMNAS CATEGORICAS {'='*40}")

df.describe(include='str')

======================================== ESTADISTICA DESCRIPTIVA COLUMNAS CATEGORICAS ========================================


,DEPARTAMENTO,MUNICIPIO,ARMAS MEDIOS,FECHA HECHO,GENERO,GRUPO ETARIO
count,660756,660752,660756,660756,658150,659732
unique,32,1021,9,5783,3,3
top,CUNDINAMARCA,Bogotá D.C. (CT),CONTUNDENTES,01/01/2020,FEMENINO,ADULTOS
freq,99054,38661,296551,360,494054,604920


**Hallazgos:**

* Las variables categoricas presentan distintos niveles de cardinalidad. `MUNICIPIO` muestra alta cardinalidad `1021 categorias`, `ARMAS MEDIOS` presenta `9 categorias`, `DEPARTAMENTO` presenta `32 categorias`, mientras que `GENERO` y `GRUPO ETARIO` presentan `3 categorias`.

* Se observa concentracion en algunas categorias dominantes, de acuerdo con la frecuencia reportada.

## 3. Analisis de Nulos:

In [87]:
# Obtencion de nulos por columna.

print(f"{'='*40} NULOS POR COLUMNA {'='*40}")

cant_nulos = df.isna().sum()
pct_nulos = (cant_nulos / len(df)*100).round(4)

reporte_nulos = pd.DataFrame({
    'nulos': cant_nulos,
    'porcentaje_(%)': pct_nulos
})

display(reporte_nulos[reporte_nulos['nulos'] > 0])

======================================== NULOS POR COLUMNA ========================================


,nulos,porcentaje_(%)
MUNICIPIO,4,0.0006
GENERO,2606,0.3944
GRUPO ETARIO,1024,0.1550


Hallazgo:

+ La columna `MUNICIPIO` presenta `4 valores nulos` (0.0006%).

+ La columna `GENERO` presenta `2606 valores nulos` (0.3944%).

+ La columna `GRUPO ETARIO` presenta `1024 valores nulos` (0.1550%).

En todos los casos, la proporción de valores nulos es `< 1%`, lo que indica una baja incidencia de datos faltantes en el dataset.

## 4. Duplicados Exactos:

In [88]:

print(f"{'='*40} DUPLICADOS EXACTOS {'='*40}")

cant_duplicados = df.duplicated().sum()
pct_duplicados = (cant_duplicados / len(df)*100).round(2)

print(f"\nLa cantidad de duplicados exactos es {cant_duplicados}, lo cual representa el {pct_duplicados}% de todo el Dataset.")

======================================== DUPLICADOS EXACTOS ========================================

La cantidad de duplicados exactos es 0, lo cual representa el 0.0% de todo el Dataset.


**Hallazgo:**

El dataset en estado crudo no presenta duplicados completos (registros identicos en todas sus columnas).

Sin embargo, esta validación debe repetirse en el cuaderno `02_limpieza.ipynb` despues de aplicar los procesos de normalizacion y estandarizacion, dado que estas transformaciones pueden revelar duplicados no evidentes en la etapa inicial (por ejemplo, diferencias en mayusculas/minusculas, tildes, formatos o espacios).

## 5. Valores Unicos Por Columna Categorica:

In [89]:
# Valores unicos por columna categorica

categorias = df.select_dtypes(include='str').columns

for columna in categorias:
    print(f"{'='*40}\n {columna} - {df[columna].nunique()} valores unicos\n{'='*40}")
    print(df[columna].value_counts(dropna=False).tail(10).to_string())
    print()

 DEPARTAMENTO - 32 valores unicos
DEPARTAMENTO
GUAJIRA       6110
PUTUMAYO      5216
ARAUCA        4792
CHOCÓ         3453
SAN ANDRÉS    1922
GUAVIARE      1807
AMAZONAS      1629
GUAINÍA        856
VICHADA        729
VAUPÉS         727

 MUNICIPIO - 1021 valores unicos
MUNICIPIO
Jordán                   7
Sipí                     7
Bojayá                   7
La Tola                  7
Recetor                  6
Abriaquí                 6
NaN                      4
San Jacinto del Cauca    4
Zapayán                  2
Cepitá                   2

 ARMAS MEDIOS - 9 valores unicos
ARMAS MEDIOS
CONTUNDENTES                   296551
SIN EMPLEO DE ARMAS            262393
NO REPORTADO                    50866
ARMA BLANCA / CORTOPUNZANTE     41619
ESCOPOLAMINA                     3663
ARMA DE FUEGO                    3044
CORTOPUNZANTES                   1155
CORTANTES                        1134
PUNZANTES                         331

 FECHA HECHO - 5783 valores unicos
FECHA HECHO
19/11/2010  

### Hallazgos:

---

#### `DEPARTAMENTO`
- Se identifican 32 departamentos unicos.  
- Consistente con la division politico-administrativa de Colombia.

---

#### `MUNICIPIO`
- Se identifican 1021 municipios unicos.  
- Se presentan 4 valores nulos.  

**Accion sugerida:**  
Revisar los registros con valores nulos y definir imputacion o exclusion segun el contexto del analisis.

---

#### `ARMAS MEDIOS`

**Hallazgo:**  
La variable presenta mezcla de categorias y subtipos (cortante, punzante, cortopunzante), generando inconsistencia en el nivel de clasificacion.

**Criterio:**  
Segun la normativa colombiana, estos subtipos pertenecen a la categoria `ARMA BLANCA`.

**Impacto:**  
Fragmentacion de categorias, aumento innecesario de cardinalidad y riesgo de interpretacion erronea.

**Accion:**  
Unificar las categorias cortante, punzante, cortopunzante y variantes en la categoria `ARMA BLANCA`.

```python
mapeo = {
    'CORTANTES': 'ARMA BLANCA',
    'PUNZANTES': 'ARMA BLANCA',
    'CORTOPUNZANTES': 'ARMA BLANCA',
    'ARMA BLANCA / CORTOPUNZANTE': 'ARMA BLANCA'
}

df['ARMAS MEDIOS'] = df['ARMAS MEDIOS'].replace(mapeo)
```

---

#### `GENERO`
- Presenta categorias: `Femenino`, `Masculino`, `No Reporta`.  
- Se identifican 2606 valores nulos.  

**Accion:**  
Imputar los valores nulos con la categoria `No Reporta`.

---

#### `GRUPO ETARIO`
- Presenta categorias: `ADULTOS`, `ADOLESCENTES`, `MENORES`.  
- Se identifican 1024 valores nulos.  

**Accion:**  
Imputar los valores nulos mediante la categoria `No Reporta` para conservar los registros y mantener consistencia semantica.

## 6. Variable `CANTIDAD`.

`CANTIDAD` es la variable numerica central del analisis, por lo cual se debe entender su distribuccion, detectar la presencia de valores extremos (outlier) y confirmar que no existen filas en 0 con valores negativos.

In [90]:
from IPython.display import display, Markdown

print(f"{'='*40} ESTADISTICA DESCRIPTIVA COLUMNA CANTIDAD {'='*40}\n")

print(df['CANTIDAD'].describe().round(2))
print(f"\nLa Moda del Dataset es: {df['CANTIDAD'].mode()[0]}")
print(f"Porcentaje de registros igual a 1: {(df['CANTIDAD'] == 1).mean()*100:.2f}%")
print(f"Registros con 0: {(df['CANTIDAD'] == 0).sum()}")
print(f"Registros negativos: {(df['CANTIDAD'] < 0).sum()}")



======================================== ESTADISTICA DESCRIPTIVA COLUMNA CANTIDAD ========================================

count    660756.00
mean          2.11
std           5.15
min           1.00
25%           1.00
50%           1.00
75%           2.00
max         164.00
Name: CANTIDAD, dtype: float64

La Moda del Dataset es: 1
Porcentaje de registros igual a 1: 72.16%
Registros con 0: 0
Registros negativos: 0


**Hallazgo:**

* El valor mas frecuente de la columna `CANTIDAD` es `1`, el cual representa el `72.16%` de los registros.

* No existenten valores en ceros y ni negativos.

* El valor maximo es `164`, lo cual no es razonable para un registro individual de victima de VIF.

## 7. Rango Temporal.

Se verifica que las fechas correspondan al periodo declarado del dataset (01/01/2010 - 31/10/2025).

In [91]:

print(f"Fecha de inicio: {pd.to_datetime(df['FECHA HECHO'], format="%d/%m/%Y", dayfirst=True, errors='coerce').min()}")
print(f"Fecha de fin: {pd.to_datetime(df['FECHA HECHO'], format="%d/%m/%Y", dayfirst=True, errors='coerce').max()}")



Fecha de inicio: 2010-01-01 00:00:00
Fecha de fin: 2025-10-31 00:00:00


**Hallazgo:**

La columna `FECHA HECHO` se encuentra en formato texto, lo que genera una evaluacion incorrecta del rango de fechas.

* **Evidencia:**  
Al aplicar la funcion `max()` sin parseo previo, la fecha maxima obtenida fue `31/12/2024`, debido a comparacion lexicografica y no cronologica.

* **Causa:**  
Las fechas no estan en tipo `datetime`, sino en tipo `object` (string).

* **Accion:**  
Se realizo el parseo de la columna a tipo `datetime` unicamente para efectos de validacion del rango de fechas, sin persistir la transformacion en el dataset.

* **Resultado:**  
Posterior al parseo, la fecha maxima corresponde al valor real y el rango de fechas es consistente.

## 8. Conclusiones De La Exploracion.

### 8.1. Resumen de Problemas de Calidad Encontrados:

| Problema | Columna(s) | Magnitud | Severidad |
|---       |---         |---       |---        |
| Duplicados exactos | Todas | 0 filas (0%) | No aplica |
| Nulos reales | GRUPO ETARIO | 1024 filas (0.1550%) | Baja |
| Nulos reales | MUNICIPIO | 4 filas (0.0006%) | Baja |
| Nulos reales | GENERO | 2606 filas (0.3944%) | Baja |
| Tipos mixtos | CODIGO DANE | — | Baja |
| Inconsistencia de clasificacion | ARMAS MEDIOS | 4 categorias solapadas | Media |
| Nombres de las columnas no normalizados (minuscula)| MUNICIPIO | 660.752 filas | Alta |
| Texto en columnas no normalizado | Todas | 660,756 | Alta |


### 8.2. Alcance Analitico del Dataset.

+ Tendencia temporal anual y mensual (2010–2025).
+ Distribucion por gonero y grupo etario.
+ Distribucion geografica por departamento y municipio.
+ Distribucion por arma/medio empleado.
+ Analisis de la variable `CANTIDAD` como metrica principal de casos registrados.
+ Cruces entre estas variables.

### 8.3. Limitaciones del Dataset.

* **Año 2025 incompleto:** el dataset contiene datos de enero a octubre de 2025 (10 meses). Las comparaciones con años anteriores deben restringirse al mismo periodo o excluir 2025 del analisis de tendencia.

* **Subregistro estructural:** los datos reflejan casos reportados a la Policia Nacional, no casos reales. La violencia intrafamiliar tiene una tasa de subregistro alta, cualquier conclusion debe mencionar esta limitacion explícitamente.

* **GENERO sin dato:** 2.606 registros (0.39%) no tienen genero registrado.

* **MUNICIPIO con formato inconsistente:** 660.752 registros presentan mezcla de mayusculas y minusculas. Sin normalizacion previa, los analisis por municipio producirian conteos fragmentados e incorrectos.

### 8.4 Hallazgos por Variable.

**`ARMAS MEDIOS`**

La variable presenta mezcla de categorías y subtipos (cortante, punzante, cortopunzante), generando inconsistencia en el nivel de clasificacion.

**Criterio:**
Segun la normativa colombiana, estos subtipos pertenecen a la categoria `ARMA BLANCA`.

**Impacto:**
Fragmentacion de categorias, aumento innecesario de cardinalidad y riesgo de interpretacion erronea en el analisis.

**Accion recomendada (para ejecutar en el notebook 02):**
- Unificar `CORTANTES`, `PUNZANTES`, `CORTOPUNZANTES` y `ARMA BLANCA / CORTOPUNZANTE` en la categoria `ARMA BLANCA` mediante un mapeo de reemplazo.
- Reemplazar `NaN` por `No Reportado`.